File: next/electron/main/pdfService.ts
import { readFile, writeFile } from 'node:fs/promises'
import { basename, dirname, extname, join, parse } from 'node:path'
import { PDFDocument } from 'pdf-lib'
import type { PdfFileInfo, PdfMergeSource, PdfOperationResult, PdfSplitTask } from '../../src/shared/contracts/pdf'
import { parsePageSelection, selectSplitPages } from '../../src/shared/utils/pageRanges'

export async function inspectPdf(path: string): Promise<PdfFileInfo> {
  ensurePdfPath(path)
  const bytes = await readFile(path)
  const document = await PDFDocument.load(bytes)
  return { path, name: basename(path), size: bytes.length, pageCount: document.getPageCount() }
}

export async function mergePdfs(sources: PdfMergeSource[], outputPath: string): Promise<PdfOperationResult> {
  if (sources.length < 2) throw new Error('Select at least two PDF files')
  const output = await PDFDocument.create()
  let pageCount = 0
  for (const source of sources) {
    ensurePdfPath(source.path)
    const document = await PDFDocument.load(await readFile(source.path))
    const pages = parsePageSelection(source.pages, document.getPageCount())
    const copied = await output.copyPages(document, pages.map((page) => page - 1))
    copied.forEach((page) => output.addPage(page))
    pageCount += copied.length
  }
  if (pageCount === 0) throw new Error('No pages selected')
  await writeFile(outputPath, await output.save())
  return { outputs: [outputPath], pageCount }
}

export async function splitPdfs(tasks: PdfSplitTask[]): Promise<PdfOperationResult> {
  if (tasks.length === 0) throw new Error('Select at least one PDF task')
  const outputs: string[] = []
  let pageCount = 0
  for (const task of tasks) {
    ensurePdfPath(task.path)
    const source = await PDFDocument.load(await readFile(task.path))
    const pages = selectSplitPages(task.pageRange, task.rule, task.customRule, source.getPageCount())
    if (pages.length === 0) throw new Error(`No pages selected for ${basename(task.path)}`)
    const output = await PDFDocument.create()
    const copied = await output.copyPages(source, pages.map((page) => page - 1))
    copied.forEach((page) => output.addPage(page))
    const parsed = parse(task.path)
    const outputPath = join(dirname(task.path), `${parsed.name.toLowerCase()}_split.pdf`)
    await writeFile(outputPath, await output.save())
    outputs.push(outputPath)
    pageCount += copied.length
  }
  return { outputs, pageCount }
}

function ensurePdfPath(path: string): void {
  if (extname(path).toLowerCase() !== '.pdf') throw new Error('Only PDF files are supported')
}

File: next-flutter/lib/features/pdf/pdf_service.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:path/path.dart' as p;

import '../../core/samples/simple_pdf.dart';
import 'page_ranges.dart';
import 'pdf_session.dart';

const maxPdfTasks = 20;

class PdfOperationResult {
  PdfOperationResult({required this.outputs, required this.pageCount});
  final List<String> outputs;
  final int pageCount;
}

PdfTaskRow inspectPdfFile(File file) {
  if (p.extension(file.path).toLowerCase() != '.pdf') {
    throw const FormatException('Only PDF files are supported');
  }
  final bytes = file.readAsBytesSync();
  final pdf = SimplePdf.parse(bytes);
  return PdfTaskRow(
    path: file.path,
    name: p.basename(file.path),
    size: bytes.length,
    pageCount: pdf.pages.length,
    pageRange: '1-${pdf.pages.length}',
  );
}

List<PdfTaskRow> appendPdfTasks(
    List<PdfTaskRow> current, List<PdfTaskRow> additions) {
  final existing = {for (final row in current) row.path};
  return [
    ...current,
    for (final row in additions)
      if (!existing.contains(row.path)) row,
  ].take(maxPdfTasks).toList();
}

PdfOperationResult splitPdfTasks(List<PdfTaskRow> tasks,
    {String? outputDirectory}) {
  if (tasks.isEmpty) throw const FormatException('Select at least one PDF task');
  final outputs = <String>[];
  var pageCount = 0;
  for (final task in tasks) {
    final source = File(task.path).readAsBytesSync();
    final pdf = SimplePdf.parse(source);
    final pages = selectSplitPages(
        task.pageRange, task.rule, task.customRule, pdf.pages.length);
    if (pages.isEmpty) {
      throw FormatException('No pages selected for ${task.name}');
    }
    final indexes = [for (final page in pages) page - 1];
    final bytes = SimplePdf.split(source, indexes);
    final directory = outputDirectory ?? p.dirname(task.path);
    Directory(directory).createSync(recursive: true);
    final name =
        '${p.basenameWithoutExtension(task.path).toLowerCase()}_split.pdf';
    final output = File(p.join(directory, name));
    output.writeAsBytesSync(bytes);
    outputs.add(output.path);
    pageCount += indexes.length;
  }
  return PdfOperationResult(outputs: outputs, pageCount: pageCount);
}

PdfOperationResult mergePdfTasks(List<PdfTaskRow> tasks, String outputPath) {
  if (tasks.length < 2) {
    throw const FormatException('Select at least two PDF files');
  }
  final documents = <Uint8List>[];
  var pageCount = 0;
  for (final task in tasks) {
    final source = File(task.path).readAsBytesSync();
    final pdf = SimplePdf.parse(source);
    final range = task.pageRange.isEmpty ? '1-${pdf.pages.length}' : task.pageRange;
    final pages = parsePageSelection(range, pdf.pages.length);
    if (pages.isEmpty) continue;
    documents.add(SimplePdf.split(source, [for (final page in pages) page - 1]));
    pageCount += pages.length;
  }
  if (pageCount == 0) throw const FormatException('No pages selected');
  final output = File(outputPath);
  output.parent.createSync(recursive: true);
  output.writeAsBytesSync(SimplePdf.merge(documents));
  return PdfOperationResult(outputs: [output.path], pageCount: pageCount);
}
